source/common/http/conn_manager_impl.cc

```diff
@@ -1099,18 +1109,25 @@ void ConnectionManagerImpl::ActiveStream::encodeData(ActiveStreamEncoderFilter*
                    end_stream);
 
   request_info_.addBytesSent(data.length());
-  response_encoder_->encodeData(data, end_stream);
-  maybeEndEncode(end_stream);
+  if (end_stream && response_trailers_) {
```

Comment:
I'm a little confused about the semantics here. When is a filter expected to add trailers for the first time? Do we want to support setting trailers during the final data frame where end_stream was otherwise true? I don't think we want to support that and we probably do? Also where is the parallel to this code on the decoding side?

Reply:
What  happens right now (since you can override trailers whenever)

1) Trailers are set during header/data when upstream has no trailers: trailers get written out at the end of `[en|de]codeData(_, true)`
2) Trailers are set during onHeader/onData when upstream HAS trailers: the original trailers override the trailers set previously (because in that case the upstream trailers are moved into `response_trailers_` AFTER we've gone through the header/data callbacks). Trailers are written out at the end of `[en|de]codeTrailers(_)`
3) Trailers are set during onTrailers: the trailers override those provided by upstream. Trailers are written out at the end of  `[en|de]codeTrailers(_)`

3) is definitely confusing, so I'd be happy to prevent that. The question to me that remains is whether we want trailers to be written out at the end of `[en|de]codeData(_, true)` or if we want setting the trailers to cause `[en|de]codeTrailers` to be executed. I think based on your question about the decoding side that you were thinking more along the lines of the latter, because iirc that's what would allow the router to include it in the upstream request?

What steered me away from that originally was that was that it doesn't seem like `addEncodedData` will cause the `onEncodedData` callbacks to run if called in `onHeaders(_, true)`.

So after thinking about it some, how about this: 
At the end of `[en|de]codeData(_, true)` we trigger `[en|de]codeTrailers` if the trailers ptr is non-null. This should cover both encoding and decoding and we'd end up writing out the trailers to the encoder like I'm already doing. If `addTrailers` is called during the ` [en|de]codeTrailers(_)` cb we can fail with an ASSERT. 

This does mean we get both a `[en|de]codeData(_, true)` and `[en|de]codeTrailers` which might be confusing, but it would allow a filter to inject trailers when it can see that there aren't already any there (because it sees `[en|de]codeData(_, true)`).

Reply:
> So after thinking about it some, how about this:
At the end of [en|de]codeData(_, true) we trigger [en|de]codeTrailers if the trailers ptr is non-null. This should cover both encoding and decoding and we'd end up writing out the trailers to the encoder like I'm already doing. If addTrailers is called during the [en|de]codeTrailers(_) cb we can fail with an ASSERT. 

I think this is more on the right track.

> This does mean we get both a [en|de]codeData(_, true) and [en|de]codeTrailers which might be confusing, but it would allow a filter to inject trailers when it can see that there aren't already any there (because it sees [en|de]codeData(_, true)).

I don't think this is going to work as it will completely confuse filters that do things when end_stream is true. I think this is the behavior you want:
1) Do what you said previously about detecting if there are trailers set during a decodeData call.
2) If trailers get set by a decode data call, make sure that subsequent filter decodeData() calls have end_stream set to false (basically key that off of whether trailers is non-null). 
3) Then dispatch trailers.

I don't think this should be too hard to implement and should cleanly cover both encode/decode. Does that make sense?

For the error checking side of things, IMO I would probably just block trailers being added in any context other than decodeData(..., true). Does it make sense anywhere else? The otehr thing to consider is what if someone calls this when a filter has been paused and then calls continue? Will it work correctly?

Reply:
Yeah that makes sense for the most part. One question:

> I would probably just block trailers being added in any context other than decodeData(..., true). 

Do you mean `decodeData(..., false)`? Allowing setting the trailers in `decodeData(..., true)` seems to contradict point 2

Reply:
No I meant what I said, that I think the only valid context to add trailers should be within a `*Data(..., true)` call or outside of a direct call in a different callback before calling `continue*()`. What I mean by (2) is that an intermediate filter can add trailers in the context of data call with end_stream true, but *subsequent* filters should then see a data call with end_stream false, followed by a trailers call. I believe this captures the intent of what your filter needs to do and IMO is pretty clear. Does that make more sense?

Reply:
Ah, that makes perfect sense. Thanks for clarifying.